mdsuite/calculators/structure_factor.py

```diff
@@ -99,246 +110,264 @@ class StructureFactor(Calculator):
-        self.radii = None
-        self.Q_arr = np.linspace(0.5, 25, 700)
 
         self.post_generation = True
 
         self.x_label = r"$$\text{Q} / nm ^{-1}$$"
         self.y_label = r"$$\text{S(Q)}$$"
         self.analysis_name = "total_structure_factor"
 
-        self.rho = None
+        self.result_series_keys = ["q", "S"]
 
-        with open_text(data, "form_fac_coeffs.csv") as file:
-            self.coeff_atomic_formfactor = pd.read_csv(
-                file, sep=","
-            )  # stores coefficients for atomic form factors
+        # Read the data from the file.
+        stream = pkg_resources.resource_stream(data.__name__, "form_fac_coeffs.csv")
+        self.form_factor_data = pd.read_csv(stream)
 
-    def __call__(self, plot=True, data_range=1):
+    @call
+    def __call__(
+        self,
+        rdf_data: Computation = None,
+        plot=True,
+        method: str = "Faber-Ziman",
+        resolution: int = 700,
+    ):
         """
         Parameters
         ----------
+        rdf_data : Computation (optional)
+                MDSuite Computation data schema from which to load the RDF data and
+                store relevant SQL meta-data information. If not give, an RDF will be
+                computed using the default RDF arguments.
         plot : bool (default=True)
-                            Decision to plot the analysis.
-        data_range : int (default=500)
-                            Range over which the property should be evaluated.
-                            This is not applicable to the current analysis as
-                            the full rdf will be calculated.
-
-        Returns
-        -------
-        None.
+                Decision to plot the analysis.
+        method : str (default=Faber-Ziman)
+                Method use to compute the weight factors.
+        resolution : int (default=700)
+                Resolution of the structure factor.
         """
+        self.plot = plot
 
-        out = {}
-        for experiment in self.experiments:
-            self.experiment = experiment
-
-            self.plot = plot
-
-            self.rho = self.experiment.number_of_atoms / (
-                self.experiment.box_array[0]
-                * self.experiment.box_array[1]
-                * self.experiment.box_array[2]
+        if isinstance(rdf_data, Computation):
+            self.rdf_data = rdf_data
+        else:
+            self.rdf_data = self.experiment.run.RadialDistributionFunction(plot=False)
+
+        # set args that will affect the computation result
+        self.args = Args(
+            number_of_bins=self.rdf_data.computation_parameter["number_of_bins"],
+            cutoff=self.rdf_data.computation_parameter["cutoff"],
+            number_of_configurations=self.rdf_data.computation_parameter[
+                "number_of_configurations"
+            ],
+            resolution=resolution,
+        )
+        self.q_values = np.linspace(0.5, 12, resolution)
```

Comment:
Out of curiosity - why is it from `0.5 -> 12`

Reply:
It is set to 12 for me to look at a plot nicely, in general it will be set to 25 as that is range over which the approximation of Gaussian basis functions on the form factors is valid. Not set to 0 to avoid an explosion.